From 2c2368bc9f393198bd8a98585c81bbcf1c2be5e5 Mon Sep 17 00:00:00 2001
From: Guy Zadicario <guy.zadicario@intel.com>
Date: Wed, 1 Apr 2020 09:04:22 +0300
Subject: [PATCH v10 15/22] misc: nnpi: Device error handling

Handle error events reported by the device. Error events are received as
event report messages - device-level messages which are not associated with
a specific channel, and therefore are handled at the driver level.

Each error type can be handled in one or more ways:
- Report the error by sending a uevent on the device (/dev/nnpi%d)
- Update a SW counter of the error type received
- Special handling of fatal errors - change device boot state and informs
  all channels.

For some of the errors, uevent notification is triggered only if its
counter has crossed a pre-defined threshold.

Signed-off-by: Guy Zadicario <guy.zadicario@intel.com>
Reviewed-by: Vaibhav Agarwal <vaibhav.agarwal@intel.com>
---
 drivers/misc/intel-nnpi/device.c | 245 ++++++++++++++++++++++++++++++++++++++-
 drivers/misc/intel-nnpi/device.h |  96 ++++++++++++++-
 2 files changed, 337 insertions(+), 4 deletions(-)

diff --git a/drivers/misc/intel-nnpi/device.c b/drivers/misc/intel-nnpi/device.c
index b9d4e84..0e5dd62 100644
--- a/drivers/misc/intel-nnpi/device.c
+++ b/drivers/misc/intel-nnpi/device.c
@@ -263,6 +263,136 @@ static void destroy_all_channels(struct nnp_device *nnpdev)
 	}
 }
 
+static void inform_device_event(struct nnp_device *nnpdev,
+				union c2h_event_report *event)
+{
+	/* 5 == largest uevent below has 4 envs + 1 NULL terminator */
+	#define MAX_ENVS 5
+	char *event_env[MAX_ENVS];
+	int num_env = 0;
+	int i;
+	bool below_threshold = false;
+
+	/*
+	 * some events needs to be informed only every time a configured
+	 * threshold.
+	 * For those events: increment the relavent counter and either
+	 * set below_threshold to indicate no inform is needed if its
+	 * value is still below the threshold or zero the counter for
+	 * the next threshold count.
+	 */
+	spin_lock(&nnpdev->lock);
+	if (event->event_code == NNP_IPC_ERROR_MCE_CORRECTABLE) {
+		nnpdev->correctable_ecc_counter++;
+		if (nnpdev->correctable_ecc_counter <
+		    nnpdev->correctable_ecc_threshold)
+			below_threshold = true;
+		else
+			nnpdev->correctable_ecc_counter = 0;
+	} else if (event->event_code == NNP_IPC_ERROR_MCE_UNCORRECTABLE &&
+		   event->event_val != 0) {
+		nnpdev->uncorrectable_ecc_counter++;
+		if (nnpdev->uncorrectable_ecc_counter <
+		    nnpdev->uncorrectable_ecc_threshold)
+			below_threshold = true;
+		else
+			nnpdev->uncorrectable_ecc_counter = 0;
+	} else if (event->event_code == NNP_IPC_ERROR_DRAM_ECC_CORRECTABLE) {
+		nnpdev->correctable_dram_ecc_counter++;
+		if (nnpdev->correctable_dram_ecc_counter <
+		    nnpdev->correctable_dram_ecc_threshold)
+			below_threshold = true;
+		else
+			nnpdev->correctable_dram_ecc_counter = 0;
+	} else if (event->event_code == NNP_IPC_CTX_DRAM_ECC_UNCORRECTABLE) {
+		nnpdev->uncorrectable_dram_ecc_counter++;
+		if (nnpdev->uncorrectable_dram_ecc_counter <
+		    nnpdev->uncorrectable_dram_ecc_threshold)
+			below_threshold = true;
+		else
+			nnpdev->uncorrectable_dram_ecc_counter = 0;
+	}
+	spin_unlock(&nnpdev->lock);
+
+	/* Do not need to inform if below threshold */
+	if (below_threshold)
+		return;
+
+#define APPEND_ENV(args...)                                            \
+({                                                                     \
+	if (num_env < MAX_ENVS)                                        \
+		event_env[num_env++] = kasprintf(GFP_KERNEL, ## args); \
+})
+
+	/*
+	 * generate uevent to inform the event
+	 */
+	if (event->event_code == NNP_IPC_ERROR_OS_CRASHED) {
+		APPEND_ENV("NNPI_EVENT=crash");
+		APPEND_ENV("NNPI_ERROR_CLASS=fatal");
+	} else if (event->event_code == NNP_IPC_ERROR_PCI_ERROR ||
+		   event->event_code == NNP_IPC_ERROR_PROTOCOL_ERROR) {
+		APPEND_ENV("NNPI_EVENT=pci_error");
+		APPEND_ENV("NNPI_ERROR_CLASS=fatal");
+	} else if (event->event_code == NNP_IPC_ERROR_MCE_CORRECTABLE) {
+		if (event->event_val != 0)
+			APPEND_ENV("NNPI_EVENT=ecc_error");
+		else
+			APPEND_ENV("NNPI_EVENT=mce_error");
+		APPEND_ENV("NNPI_ERROR_CLASS=correctable");
+	} else if (event->event_code == NNP_IPC_ERROR_MCE_UNCORRECTABLE) {
+		if (event->event_val != 0)
+			APPEND_ENV("NNPI_EVENT=ecc_error");
+		else
+			APPEND_ENV("NNPI_EVENT=mce_error");
+		APPEND_ENV("NNPI_ERROR_CLASS=uncorrectable");
+	} else if (event->event_code == NNP_IPC_ERROR_MCE_UNCORRECTABLE_FATAL) {
+		if (event->event_val != 0)
+			APPEND_ENV("NNPI_EVENT=ecc_error");
+		else
+			APPEND_ENV("NNPI_EVENT=mce_error");
+		APPEND_ENV("NNPI_ERROR_CLASS=fatal");
+	} else if (event->event_code == NNP_IPC_THERMAL_TRIP_EVENT) {
+		APPEND_ENV("NNPI_EVENT=thermal");
+		APPEND_ENV("NNPI_ERROR_CLASS=no_error");
+		APPEND_ENV("NNPI_THERMAL_TRIP=%d", event->event_val);
+		APPEND_ENV("NNPI_THERMAL_TRIP_DIR=%s",
+			   (event->obj_id_2 >= event->obj_id ?  "up" : "down"));
+	} else if (event->event_code == NNP_IPC_ERROR_DRAM_ECC_CORRECTABLE) {
+		APPEND_ENV("NNPI_EVENT=dram_ecc_error");
+		APPEND_ENV("NNPI_ERROR_CLASS=correctable");
+	} else if (event->event_code == NNP_IPC_CTX_DRAM_ECC_UNCORRECTABLE) {
+		APPEND_ENV("NNPI_EVENT=dram_ecc_error");
+		APPEND_ENV("NNPI_ERROR_CLASS=uncorrectable");
+	} else if (event->event_code ==
+		   NNP_IPC_ERROR_DRAM_ECC_UNCORRECTABLE_FATAL) {
+		APPEND_ENV("NNPI_EVENT=dram_ecc_error");
+		APPEND_ENV("NNPI_ERROR_CLASS=fatal");
+	} else if (event->event_code == NNP_IPC_DMA_HANG_DETECTED) {
+		APPEND_ENV("NNPI_EVENT=dma_hang_error");
+		APPEND_ENV("NNPI_ERROR_CLASS=uncorrectable");
+	} else if (event->event_code == NNP_IPC_FATAL_DMA_HANG_DETECTED) {
+		APPEND_ENV("NNPI_EVENT=dma_hang_error");
+		APPEND_ENV("NNPI_ERROR_CLASS=fatal");
+	}
+
+	if (num_env > 0) {
+		/* check that all kasprintf succeeded */
+		for (i = 0; i < num_env; i++)
+			if (!event_env[i])
+				break;
+
+		if (i == num_env && num_env < MAX_ENVS) {
+			event_env[num_env] = NULL;
+			kobject_uevent_env(&nnpdev->dev->kobj, KOBJ_CHANGE,
+					   event_env);
+		}
+
+		for (i = 0; i < num_env; ++i)
+			kfree(event_env[i]);
+	}
+}
+
 void nnpdev_submit_device_event_to_channels(struct nnp_device *nnpdev,
 					    union c2h_event_report *event_msg,
 					    bool force)
@@ -372,6 +502,54 @@ static void handle_channel_destroy(struct nnp_device *nnpdev,
 	nnp_chan_put(cmd_chan);
 }
 
+static void handle_device_critical_error(struct nnp_device *nnpdev,
+					 union c2h_event_report *event_msg)
+{
+	spin_lock(&nnpdev->lock);
+	switch (event_msg->event_code) {
+	case NNP_IPC_ERROR_OS_CRASHED:
+		nnpdev->state |= NNP_DEVICE_KERNEL_CRASH;
+		nnpdev->host_crash_dump.dump_size =
+				((u32)event_msg->obj_id_2 << 16) |
+				(u32)event_msg->obj_id;
+		nnpdev->counters.uncorr.os_crashed++;
+		break;
+	case NNP_IPC_ERROR_PROTOCOL_ERROR:
+		nnpdev->state |= NNP_DEVICE_PROTOCOL_ERROR;
+		break;
+	case NNP_IPC_ERROR_MCE_UNCORRECTABLE_FATAL:
+		nnpdev->state |= NNP_DEVICE_FATAL_MCE_ERROR;
+		nnpdev->counters.uncorr.os_crashed++;
+		if (event_msg->event_val == 1)
+			nnpdev->counters.uncorr.ecc_fatal++;
+		else
+			nnpdev->counters.uncorr.mce_fatal++;
+		break;
+	case NNP_IPC_ERROR_DRAM_ECC_UNCORRECTABLE_FATAL:
+		nnpdev->state |= NNP_DEVICE_FATAL_DRAM_ECC_ERROR;
+		nnpdev->counters.uncorr.dram_ecc_fatal++;
+		break;
+	case NNP_IPC_ERROR_FATAL_ICE_ERROR:
+		nnpdev->state |= NNP_DEVICE_FATAL_ICE_ERROR;
+		break;
+	case NNP_IPC_ERROR_CARD_RESET:
+		if (nnpdev->state & NNP_DEVICE_HANG)
+			dev_info(nnpdev->dev,
+				 "hang detected - reset is needed\n");
+		break;
+	case NNP_IPC_FATAL_DMA_HANG_DETECTED:
+		nnpdev->counters.uncorr.dma_hang_fatal++;
+		nnpdev->state |= NNP_DEVICE_FATAL_DMA_ERROR;
+		break;
+	default:
+		dev_err(nnpdev->dev, "Unknown event received - %u\n",
+			event_msg->event_code);
+	}
+	spin_unlock(&nnpdev->lock);
+
+	nnpdev_submit_device_event_to_channels(nnpdev, event_msg, true);
+}
+
 /*
  * this function handle device-level event report message.
  * which is usually affect the entire device and not a single channel
@@ -380,9 +558,33 @@ static void process_device_event(struct nnp_device *nnpdev,
 				 union c2h_event_report *event_msg)
 {
 	if (is_card_fatal_event(event_msg->event_code)) {
-		nnpdev_submit_device_event_to_channels(nnpdev, event_msg, true);
+		/* Handle device critical error */
+		handle_device_critical_error(nnpdev, event_msg);
 	} else {
 		switch (event_msg->event_code) {
+		case NNP_IPC_ERROR_MCE_CORRECTABLE:
+			nnpdev->counters.corr.ecc++;
+			break;
+		case NNP_IPC_ERROR_DRAM_ECC_CORRECTABLE:
+			nnpdev->counters.corr.dram_ecc++;
+			break;
+		case NNP_IPC_ERROR_MCE_UNCORRECTABLE:
+			dev_dbg(nnpdev->hw_dev->dev,
+				"ECC error received code - %u\n",
+				event_msg->event_code);
+			if (event_msg->event_val == 1)
+				nnpdev->counters.uncorr.ecc_nonfatal++;
+			else
+				nnpdev->counters.uncorr.mce_nonfatal++;
+			break;
+		case NNP_IPC_CTX_DRAM_ECC_UNCORRECTABLE:
+			nnpdev->counters.uncorr.dram_ecc_nonfatal++;
+			break;
+		case NNP_IPC_THERMAL_TRIP_EVENT:
+			dev_dbg(nnpdev->hw_dev->dev,
+				"Thermal trip event num=%d temp=%u\n",
+				event_msg->event_val, event_msg->obj_id_2);
+			break;
 		case NNP_IPC_CREATE_CHANNEL_SUCCESS:
 		case NNP_IPC_CREATE_CHANNEL_FAILED:
 		case NNP_IPC_CHANNEL_SET_RB_SUCCESS:
@@ -410,6 +612,10 @@ static void process_device_event(struct nnp_device *nnpdev,
 		case NNP_IPC_CHANNEL_UNMAP_HOSTRES_SUCCESS:
 			handle_channel_unmap_hostres(nnpdev, event_msg);
 			break;
+		case NNP_IPC_DMA_HANG_DETECTED:
+			nnpdev->counters.uncorr.dma_hang_nonfatal++;
+			break;
+
 		default:
 			dev_err(nnpdev->hw_dev->dev,
 				"Unknown event received - %u\n",
@@ -419,6 +625,32 @@ static void process_device_event(struct nnp_device *nnpdev,
 
 		nnpdev_submit_device_event_to_channels(nnpdev, event_msg, false);
 	}
+
+	/*
+	 * ECC errors may be channel specific - call to notify channel
+	 */
+	if (event_msg->obj_valid &&
+	    event_msg->event_code == NNP_IPC_CTX_DRAM_ECC_UNCORRECTABLE) {
+		union c2h_event_report ev;
+		struct nnp_chan *cmd_chan;
+
+		ev.value = event_msg->value;
+		ev.chan_id = event_msg->obj_id;
+		ev.chan_valid = 1;
+
+		cmd_chan = nnpdev_find_channel(nnpdev, ev.chan_id);
+		if (cmd_chan) {
+			nnp_chan_add_response(cmd_chan, (u64 *)&ev, sizeof(ev));
+			nnp_chan_put(cmd_chan);
+		} else {
+			dev_err(nnpdev->hw_dev->dev,
+				"Got channel severity DRAM ECC error for non existing channel id %d!!\n",
+				ev.chan_id);
+		}
+	}
+
+	/* inform device event */
+	inform_device_event(nnpdev, event_msg);
 }
 
 struct event_report_work {
@@ -658,8 +890,17 @@ void nnpdev_process_messages(struct nnp_device *nnpdev, u64 *hw_msg,
 		j += msg_size;
 	} while (j < nof_msg);
 
-	if (fatal_protocol_error)
+	if (fatal_protocol_error) {
+		u64 ev;
+
+		ev = FIELD_PREP(NNP_C2H_OP_MASK, NNP_IPC_C2H_OP_EVENT_REPORT);
+		ev |= FIELD_PREP(NNP_C2H_EVENT_REPORT_CODE_MASK,
+				 NNP_IPC_ERROR_PROTOCOL_ERROR);
+
 		nnpdev->state |= NNP_DEVICE_PROTOCOL_ERROR;
+		handle_event_report(nnpdev, &ev, 1);
+		j = hw_nof_msg;
+	}
 
 	/*
 	 * if unprocessed messages left, copy it to the pending messages buffer
diff --git a/drivers/misc/intel-nnpi/device.h b/drivers/misc/intel-nnpi/device.h
index 6b7b5a6..a81b11b 100644
--- a/drivers/misc/intel-nnpi/device.h
+++ b/drivers/misc/intel-nnpi/device.h
@@ -44,6 +44,7 @@
 #define NNP_DEVICE_FAILED_VERSION    BIT(16)
 #define NNP_DEVICE_BOOT_FAILED       BIT(17)
 #define NNP_DEVICE_HOST_DRIVER_ERROR BIT(18)
+#define NNP_DEVICE_FATAL_DMA_ERROR   BIT(19)
 #define NNP_DEVICE_KERNEL_CRASH	     BIT(20)
 #define NNP_DEVICE_PCI_ERROR         BIT(21)
 #define NNP_DEVICE_CARD_IN_RESET     BIT(22)
@@ -74,6 +75,72 @@ struct query_version_work {
 	bool running;
 };
 
+struct nnp_device_counters {
+	struct {
+		/*
+		 * Number of times device needed to be reset
+		 * due to device fatal error
+		 */
+		unsigned long  os_crashed;
+		/*
+		 * Number of times a non-fatal
+		 * uncorrectable ECC error happened
+		 * on device
+		 */
+		unsigned long  ecc_nonfatal;
+		/*
+		 * Number of times a fatal, uncorrectable
+		 * ECC error happened on device
+		 */
+		unsigned long  ecc_fatal;
+		/*
+		 * Number of times a non-fatal
+		 * uncorrectable ECC error happened
+		 * on device DRAM
+		 */
+		unsigned long  dram_ecc_nonfatal;
+		/*
+		 * Number of times a fatal, uncorrectable
+		 * ECC error happened on device DRAM
+		 */
+		unsigned long  dram_ecc_fatal;
+		/*
+		 * Number of times a non-fatal
+		 * uncorrectable MCE error happened
+		 * on device
+		 */
+		unsigned long  mce_nonfatal;
+		/*
+		 * Number of times a fatal, uncorrectable MCE
+		 * error happened on device
+		 */
+		unsigned long  mce_fatal;
+		/*
+		 * Number of times device DMA engine got hanged
+		 * and recovered.
+		 */
+		unsigned long  dma_hang_nonfatal;
+		/*
+		 * Number of times device DMA engine got hanged
+		 * and could not be recovered.
+		 */
+		unsigned long  dma_hang_fatal;
+	} uncorr;
+
+	struct {
+		/*
+		 * Number of times a correctable ECC error
+		 * happened on device
+		 */
+		unsigned long  ecc;
+		/*
+		 * Number of times a correctable ECC error
+		 * happened on device DRAM
+		 */
+		unsigned long  dram_ecc;
+	} corr;
+};
+
 /**
  * union c2h_event_report - packs device event report response in 64-bit object
  * @event_code: device event code, indicate which event has happened
@@ -122,7 +189,8 @@ struct query_version_work {
  * @wq: singlethread workqueue for processing device's response messages.
  * @image_wait_node: list node to include this device is a list of devices
  *                   waiting for the same boot image to be loaded from disk.
- * @lock: protects accesses to @state, @inbound_mem and @cmd_chan_hash
+ * @lock: protects accesses to @state, @inbound_mem, @cmd_chan_hash and
+ *        ecc counters and thresholds.
  * @is_recovery_bios: true if device has booted from the recovery bios flash
  * @boot_image_loaded: true if boot image load has started
  * @reset_boot_image_path: boot image to load on next device reset.
@@ -154,6 +222,19 @@ struct query_version_work {
  * @ipc_chan_resp_op_size: holds response size for each possible channel
  *                         response.
  * @ipc_chan_cmd_op_size: holds command size for each possible channel command.
+ * @counters: counter values for various device events
+ * @correctable_ecc_threshold: threshold value of error counter for generating
+ *                             uevent informing on correctable ecc error.
+ * @correctable_ecc_counter: correctable ecc error counter
+ * @uncorrectable_ecc_threshold: threshold error counter for generating
+ *                               uevent informing on uncorr ecc error.
+ * @uncorrectable_ecc_counter: uncorrectable ecc error counter
+ * @correctable_dram_ecc_threshold: threshold error counter for generating
+ *                                  uevent informing on corr dram ecc error.
+ * @correctable_dram_ecc_counter: correctable dram ecc error counter
+ * @uncorrectable_dram_ecc_threshold: threshold error counter for generating
+ *                                    uevent informing on uncorr dram ecc error
+ * @uncorrectable_dram_ecc_counter: uncorrectable dram ecc error counter
  */
 struct nnp_device {
 	struct kref            ref;
@@ -171,7 +252,7 @@ struct nnp_device {
 	struct list_head        image_wait_node;
 	/*
 	 * protects accesses to @state, @inbound_mem,
-	 * @cmd_chan_hash
+	 * @cmd_chan_hash and ecc counters and thresholds
 	 */
 	spinlock_t     lock;
 	bool           is_recovery_bios;
@@ -206,8 +287,19 @@ struct nnp_device {
 
 	struct cdev cdev;
 	struct device *dev;
+
 	u8   ipc_chan_resp_op_size[32];
 	u8   ipc_chan_cmd_op_size[32];
+
+	struct nnp_device_counters counters;
+	unsigned long  correctable_ecc_threshold;
+	unsigned long  correctable_ecc_counter;
+	unsigned long  uncorrectable_ecc_threshold;
+	unsigned long  uncorrectable_ecc_counter;
+	unsigned long  correctable_dram_ecc_threshold;
+	unsigned long  correctable_dram_ecc_counter;
+	unsigned long  uncorrectable_dram_ecc_threshold;
+	unsigned long  uncorrectable_dram_ecc_counter;
 };
 
 void nnpdev_module_init(void);
-- 
1.8.3.1

